mlflow_elasticsearchstore/elasticsearch_store.py

```diff
@@ -247,18 +247,41 @@ def get_metric_history(self, run_id: str, metric_key: str) -> List[Metric]:
                  response["hits"]["hits"][0].inner_hits.metrics.hits.hits]
                 if (len(response["hits"]["hits"]) != 0) else [])
 
-    def list_all_columns(self, experiment_id: str, run_view_type: str) -> Columns:
-        stages = LifecycleStage.view_type_to_stages(run_view_type)
+    def _list_columns(self, experiment_id: str, stages: List[LifecycleStage],
+                      column_type: str, columns: List[str],
+                      after_key: str = None, size: int = 100) -> None:
         s = Search(index="mlflow-runs").filter("match", experiment_id=experiment_id) \
             .filter("terms", lifecycle_stage=stages)
-        for col in ['latest_metrics', 'params', 'tags']:
-            s.aggs.bucket(col, 'nested', path=col)\
-                .bucket(f'{col}_keys', "terms", field=f'{col}.key')
-        response = s.execute()
-        metrics = [m.key for m in response.aggregations.latest_metrics.latest_metrics_keys.buckets]
-        params = [p.key for p in response.aggregations.params.params_keys.buckets]
-        tags = [t.key for t in response.aggregations.tags.tags_keys.buckets]
-        return Columns(metrics=metrics, params=params, tags=tags)
+        if after_key is None:
+            s.aggs.bucket(column_type, 'nested', path=column_type) \
+                .bucket(f'{column_type}_keys', "composite", size=size,
+                        sources=[{"key": {"terms": {"field": f'{column_type}.key'}}}])
+        else:
+            s.aggs.bucket(column_type, 'nested', path=column_type) \
+                .bucket(f'{column_type}_keys', "composite", size=size,
+                        sources=[{"key": {"terms": {"field": f'{column_type}.key'}}}],
+                        after={"key": after_key})
+        response = s.source(False).execute()
+        new_columns = [column.key.key for column in attrgetter(
+            f'aggregations.{column_type}.{column_type}_keys.buckets')(response)]
+        columns += new_columns
+        if len(new_columns) == size:
+            after_key = attrgetter(
+                f'aggregations.{column_type}.{column_type}_keys.after_key.key')(response)
+            return self._list_columns(experiment_id, stages, column_type, columns, after_key)
```

Comment:
You don't need to use recursive calls here, you can do everything with a while

Reply:
Ok, thought recursive was cleaner